### PR TITLE
feat(goleveldb,rocksdb): optimization: remove redundant check from iterator

### PR DESCRIPTION
**Pull request by faddat**
This is a cherry pick from terra's performance oriented branch.

It removes an unnecessary check when we use the iterator.

### CODE REVIEW

**Comment by faddat** [comment]
@creachadair I really couldn't say for sure if this one is safe.

**Comment by 08d2** [comment]
> It removes an unnecessary check when we use the iterator.

How do you know it is unnecessary?

**Comment by faddat** [comment]
@08d2 I don't for sure, which is why I said

```

@creachadair I really couldn't say for sure if this one is safe.

```

**Comment by 08d2** [comment]
> @08d2 I don't for sure, which is why I said
> 
> ```
> 
> @creachadair I really couldn't say for sure if this one is safe.
> ```

Then I suppose this PR should be closed, no?

**Comment by faddat** [comment]
None of us should be sure.

My preference: I change this to a draft, and figure out if it's safe.

**Comment by faddat** [comment]
@creachadair on this, I'd kindly request aid in figuring out how to make the certain determination that this code is safe.  It's part of the omnibus upgrade branch.  My comments about safety are driven by disabling these tests:

**Comment by 08d2** [comment]
@faddat You have removed a fair amount of code which verifies certain invariants during tests. If that code was added again, would the corresponding tests fail? If they would fail, can you explain why those failures are safe to ignore?

**Comment by faddat** [comment]
Hi @08d2 -- I have mainly experiential info on this.  Basically, those iterator checks that are removed here, are what are required to pass the the tests that I removed here.

"it works".

Now, if you're able to help come up with some new tests, that would be super useful.  For both this and #237 -- the reality is that I'm very literally just doing what I can observe to work well. 

I would like to improve testing, and restructure this repo, and ensure that we remove badgerdb from future timelines for both the cosmos-sdk and tendermint -- or have very good reason for using it.

**Comment by 08d2** [comment]
But why did you remove those tests?

**Comment by faddat** [comment]
ah-- look at the corresponding code

**Comment by 08d2** [comment]
Looking at the code explains the "what" but not the "why" — can you ELI5? 😇

**Comment by faddat** [comment]
sure.   The tests were dependent on the "redundant check from iterator".

**Comment by faddat** [comment]
https://github.com/tendermint/tm-db/pull/250/files#diff-61cc79a7f545ad48e7bce30b84f13aa4cfd4250b9898f0f6a18ccb38984acf10


See checknextpanics?

That would return the specific panic that the tests were looking for.

**Comment by 08d2** [comment]
I'm sorry, I don't follow. I'd appreciate it if you could take a few minutes to write out exactly what this PR is doing, and why you believe it is (or could be) safe. Connect the dots for me :) because at the moment, it's not obvious to me why any of these changes have been made, nor that any of them are actually safe.

**Comment by faddat** [comment]
* https://github.com/tendermint/tm-db/pull/280

-- anyhow @08d2 -- I'd prefer to keep this as a draft, walk you through reasoning too. 

This PR, which is also contained in #237 -- is based on some work from Terra on performance, you can find it at their tm-db fork:

https://github.com/terra-money/tm-db

Unfortunately, they weren't super insistent on the tests passing there.  

1) it's possible that this PR is entirely in error eg: maybe their code wasn't safe, either.  
2) Seeking help to either:

* "this is plain usafe, close"
* make better tests
